### PR TITLE
S3UTILS-58 diff stream without digest checks

### PR DESCRIPTION
**Pull request by jonathan-gramain**
Output differences between the input stream, consisting of
{ key: string, value: string } objects sorted by key, and a bucketd
interface that can be queried via HTTP for ranges of keys from
given buckets.

Input items are meant to be coming from a raw leveldb database
listing, transformed via a DBListStream object to be comparable with
bucketd listing.

### CODE REVIEW

File: CompareRaftMembers/DiffStream.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ */
+class DiffStream extends stream.Transform {
+    /**
+     * @constructor
+     * @param {object} params - constructor params
+     * @param {string} params.bucketdHost - bucketd host name or IP address
+     * @param {number} params.bucketdPort - bucketd API port
+     * @param {number} [params.maxBufferSize=1000] - maximum number of
+     * items to bufferize
+     * @param {class} [params.BucketStreamClass=BucketStream] -
```
Comment (watsaqat) [inline CompareRaftMembers/DiffStream.js:25]: why is this needed in the params instead of modifying BucketStreamClass after instantiation in the tests?

File: tests/unit/CompareRaftMembers/DiffStream.js [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+                                marker: null,
+                                lastKey: null,
+                            },
+                        ],
+                    },
+                ],
+            },
+            {
+                desc: 'with 7777 identical entries in db and bucketd',
+                dbContents: () => {
```
Comment (watsaqat) [inline tests/unit/CompareRaftMembers/DiffStream.js]: you can use getters https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get so you don't have to use explicit checks such as
`if (typeof dbContents === 'function') { .. } else { .. }`

File: tests/unit/CompareRaftMembers/DiffStream.js [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+                        .on('data', data => {
+                            output.push(data);
+                        })
+                        .on('end', () => {
+                            expect(output).toEqual(expectedOutput);
+                            expect(MOCK_BUCKET_STREAM_REQUESTS_MADE)
+                                .toEqual(withDigests.expectedBucketStreamRequests);
+                            done();
+                        })
+                        .on('error', done);
```
Comment (watsaqat) [inline tests/unit/CompareRaftMembers/DiffStream.js:533]: if an error occurs before the 'end' event, wouldn't the test appear to succeed here?

File: CompareRaftMembers/DiffStream.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        if (this.inputBuffer.length > 0) {
+            this._compareInputBufferWithBucketd(null, callback);
+        } else {
+            callback();
+        }
+    }
+
+    _parseItem(item) {
+        const { key: fullKey, value } = item;
+        const slashIndex = fullKey.indexOf('/');
```
Comment (watsaqat) [inline CompareRaftMembers/DiffStream.js]: want to make sure that bucketName can't be suffixed with '/' for some reason as that would mess up this split.

**Comment by watsaqat** [comment]
could you add some JSDocs to the DiffStream methods?

File: CompareRaftMembers/DiffStream.js (hunk 4 above)
Comment (jonathan-gramain) [reply CompareRaftMembers/DiffStream.js]: yes, bucket names are DNS-compatible, hence no slash or other forbidden character

File: tests/unit/CompareRaftMembers/DiffStream.js (hunk 3 above)
Comment (jonathan-gramain) [reply tests/unit/CompareRaftMembers/DiffStream.js:533]: no, since the error event will be passed an error argument, it will automatically be passed to the `done` callback, then failing the test.

File: CompareRaftMembers/DiffStream.js (hunk 1 above)
Comment (jonathan-gramain) [reply CompareRaftMembers/DiffStream.js:25]: It's because the instanciation happens internally, potentially multiple times, from the `_compareInputBufferWithBucketd` helper.

**Review by tmacro — APPROVED** [review]

**Review by watsaqat — APPROVED** [review]

**Comment by jonathan-gramain** [comment]
@watsaqat I added JSDoc, I also reorganized the code a bit and squashed
